### ISSUE TITLE
Provide better protection against manually installed requirements

### ISSUE THREAD

**Issue by freakboy3742** [beeware/briefcase#431]
A frequently reported error: A user creates a venv, manually installs a package into the venv, and runs `briefcase dev`. The app runs succcessfully. The user then runs `briefcase run`, and the app fails because the package can't be found.

This is 100% working-as-intended; the package that was installed manually needs to be added to the `requires` list to be packaged with the independent environment used in the running app.

However, it's sufficiently confusing that we should provided better protection against this mode of failure. 

Two possible mitigations:

1) Don't use the local venv when running `briefcase dev`. Tox, for example, creates an independent venv and runs tests in that venv. If `briefcase dev` used a similar approach, the environment where the app runs would be isolated from the venv where briefcase is running. This would have the added advantage that development tooling (like briefcase) wouldn't be present in the `dev` environment.

2) Raise a warning during `briefcase run` if there are packages in the development environment that aren't present in the `requires` list.  

The downside to (1) is a potentially slower development cycle, as starting the app will take a little more time (as the venv needs to be created/activated before the app runs)

The downside to (2) is the noise associated with packages that are in the dev environment legitimately - e.g., briefcase itself. It may be necessary to have a 'known dev packages' list in pyproject.toml to silence this warning.

**Comment by jgirardet** [comment]
My answer is not 100% about it but it's the same idea.
We don't have yet a way to install dev packages. It can be done using `briefcase dev -d` but will run the app too.
A `briefcase dev -i` would install  all the depencies from pyrproject `requires`.
A require-dev in pyproject would solve the probleme of other dependencies (pytest, black,....) it's basically how poetry works and it's really straightforward.

About the real question (is it : "how to protect programmers from themselves?")  I'm not sure it's in the scope of briefcase.  Check every subdependency would be big work I think. Does it worth it ?.
And I don't mention the case where some one use a "dev" dependency in runtime code but forgot it.
Maybe should it be solved in #399 

If at the bare minimum we can have **one** source for dependencies, it's good.

**Comment by freakboy3742** [comment]
I completely agree that having a mechanism to install dependencies is something worth adding, if only for testing. However, I'm not sure it addresses the problem I described.

The use case that is a problem is:
 * User decides to use `requests` in their app
 * User run `pip install requests`, and adds `import requests` to their app code
 * User runs `briefcase dev`; app runs fine.
 * User runs `briefcase run`; app fails because `requests` *isn't* in the `requires` list in `pyproject.toml`.

The issue is that we can't prevent people from pip installing packages (nor should we want to); but if they do, they might not realise there's a parallel action required (adding to `pyproject.toml`)

Having an "install dev requirements" option means there's a new way to install that avoids the problem; but we won't ever prevent the muscle memory of users doing a pip install directly. That's the use case we need to detect.

Having a clear list of "development requirements" would help in reverse engineering "what is in your environment that isn't in your requires list" - but even then, the list is likely to be incomplete. For example, `ipython` will be a common addition in many local environments, but it shouldn't ever be in a development requirements list (you definitely don't need ipython for CI).

**Comment by freakboy3742** [comment]
I'm going to close this; Briefcase 0.4.0 now uses isolated virtual environments for the dev command, so unless a dependency is listed in `requires`, it won't be used *at all*, so this problem has been mitigated.
